### ISSUE TITLE
False positive for `no-document-import-in-page`

### ISSUE THREAD

**Issue by GrantGryczan** [vercel/next.js#28596]
### What version of Next.js are you using?

11.1.0

### What version of Node.js are you using?

14.17.1

### What browser are you using?

Chrome

### What operating system are you using?

Windows

### How are you deploying your application?

next start

### Describe the Bug

The `@next/next/no-document-import-in-page` ESLint rule displays an error in a custom `_document` if it is located at `pages/_document/index.js`, `pages/_document/index.jsx`, `pages/_document/index.ts`, or `pages/_document/index.tsx`.

### Expected Behavior

I expect there to be no ESLint error from the `@next/next/no-document-import-in-page` rule in all valid custom `_document` file locations.

### To Reproduce

Create `pages/_document/index.jsx` as follows with the ESLint rule enabled.

```jsx
import Document, { Html, Head, Main, NextScript } from 'next/document';

class MyDocument extends Document {
	render() {
		return (
			<Html lang="en">
				<Head />
				<body>
					<Main />
					<NextScript />
				</body>
			</Html>
		);
	}
}

export default MyDocument;
```

This results in an error on line 1.

**Comment by kimbaudi** [comment]
I'm getting the same eslint error after updating next.js from 11.1.0 to 11.1.1. I don't recall getting this error when I was on next.js 11.1.0. In my case, the custom `_document.tsx` file is located under `src/pages`.

other npm packages I updated that might cause this error:
- @types/node (from 16.7.5 to 16.7.7)
- @typescript-eslint/eslint-plugin (from 4.29.3 to 4.30.0)
- @typescript-eslint/parser (from 4.29.3 to 4.30.0)
- eslint-config-next (from 11.1.0 to 11.1.1)
- eslint-plugin-react (from 7.24.0 to 7.25.1)

**Comment by JamesSingleton** [comment]
Yeah I just started getting this when I went to update to next 11.1.1. It doesn't like/consider `_document.tsx` as a valid file.

**Comment by kimbaudi** [comment]
this is not just a problem for `_document.tsx`, but also `_app.tsx`. reverting to `next` & `eslint-config-next` v 11.1.0 until this is resolved.

I wonder if this is caused by the recent typescript version update and eslint and/or typescript @types not yet being updated with the latest changes to typescript. 🤷‍♂️

**Comment by samsisle** [comment]
Is there a reason why you are creating a `_document` directory instead of `pages/_document.tsx`?

**Comment by JamesSingleton** [comment]
Although I do agree that creating a `_document` directory is incorrect, @kimbaudi is experiencing the same issue and is not creating a `_document` directory.

**Comment by GrantGryczan** [comment]
> Is there a reason why you are creating a `_document` directory instead of `pages/_document.tsx`?

It shouldn't matter--this issue is a bug no matter what I'm doing it for, because it is a valid location.

Regardless, my personal (and again, irrelevant) reason is just because it's more consistent, since every other page and component in my project uses a folder (usually in order to be able to have a `styles.module.scss` in the same folder), including `pages/_error` for example.

**Comment by GrantGryczan** [comment]
> Is there a reason why you are creating a `_document` directory instead of `pages/_document.tsx`?

@samsisle It shouldn't matter--this issue is a bug no matter what reason I'm doing it for, because it is a valid location.

Regardless, my personal (and again, irrelevant) reason is just because it's more consistent, since every other page and component in my project uses a folder (usually in order to be able to have a `styles.module.scss` in the same folder), including `pages/_error` for example.

**Comment by rtritto** [comment]
Same problem here.
I'm using on Windows yarn v3, next 11.1.2 and typescript 4.4.2.

My updates diff:
- @types/node (from 16.7.6 to 16.7.8)
- @typescript-eslint/eslint-plugin (from 4.29.3 to 4.30.0)
- @typescript-eslint/parser (from 4.29.3 to 4.30.0)
- eslint-config-next (11.1.2)
- eslint-plugin-react (from 7.25.0 to 7.25.1)

**Comment by alex-drocks** [comment]
I get the same false positive.

All I did was update from Next v1.1.0 to Next 1.1.2 and update eslint-config-next from 11.1.0 to 11.1.2.

I'm on Windows 10, using node v16.7.0.
Everything was fine with next 1.1.0

**Comment by rtritto** [comment]
It is introduced by [v11.1.1-canary.17](https://github.com/vercel/next.js/releases/tag/v11.1.1-canary.17) with:
- [ESLint Plugin] Handles edge case for `no-import-document-in-page` rule: #28261

No problems with v11.1.1-canary.16.

**Comment by mkayander** [comment]
I got this issue at Next version `11.0.1`. I see that my `@next/eslint-plugin-next` package updated automatically by Yarn from `11.1.0` to `11.1.2`. So i assume this is the cause. Added ignore line comment temporaly.

**Comment by housseindjirdeh** [comment]
Thanks for logging this issue @GrantGryczan, and thanks to everyone else sharing their details which always helps. For some reason, I can't reproduce this. I've tried next & eslint-config-next versions 11.1.0, 11.1.1, and 11.1.2 and it seems to work fine . I've even tried spinning up a new application and still couldn't reproduce it:

1. `npx create-next-app test-next-app`
2. `cd test-next-app`
3. Verify versions in `package.json` (both `next` & `eslint-config-next` are versions 11.1.2)
4. Create `pages/_document/index.jsx` with a custom document:

```
import Document, { Html, Head, Main, NextScript } from 'next/document'

class MyDocument extends Document {
  render() {
    return (
      <Html>
        <Head />
        <body>
          <Main />
          <NextScript />
        </body>
      </Html>
    )
  }
}

export default MyDocument
```

5. Run `yarn lint`
6. "No ESLint warnings or errors" 🤔 

I've also tried variations of the document file (`pages/_document/index.jsx`, `src/pages/_document/index.js`, `pages/_document.tsx`) and still not seeing the issue. 

> It is introduced by v11.1.1-canary.17 with:
> 
> [ESLint Plugin] Handles edge case for no-import-document-in-page rule: [ESLint Plugin] Handles edge case for no-import-document-in-page rule #28261
>
> No problems with v11.1.1-canary.16.

That's the only major change that's been introduced to the file since 11.1.0, but doesn't seem to be causing the error on my machine. I can revert the changes if necessary though.

Can anyone share their repository of where they're seeing this issue if it's public? Not sure if I'm missing something obvious here 😅 . Also, I would suggest running `yarn lint --no-cache` in case caching is causing this wonky behavior, so if anyone finds that happens to resolve this issue please let me know.

**Comment by kimbaudi** [comment]
@housseindjirdeh - I think you need to create a TypeScript project with create-next-app (i.e., `npx create-next-app test-next-app --ts`). Let me see if I can recreate the issue.

**Comment by kimbaudi** [comment]
@housseindjirdeh - I've created a minimal reproducible repo at https://github.com/kimbaudi/test-next-app that results in the error. If you look at `package.json`, you will see that it is using `next` and `eslint-config-next` v11.1.2. If you downgrade to v11.1.0, you won't get the error. I believe upgrading to v11.1.1 and v11.1.2 produces the error.

**Comment by rtritto** [comment]
Steps to reproduce:
1. `yarn init -y`
2. `yarn set version berry`
3. `yarn add eslint-config-next next react react-dom eslint`
4. Create `pages/_document.jsx` or `pages/_document/index.jsx`
```js
import Document, { Html, Head, Main, NextScript } from 'next/document'

class MyDocument extends Document {
  render() {
    return (
      <Html>
        <Head />
        <body>
          <Main />
          <NextScript />
        </body>
      </Html>
    )
  }
}

export default MyDocument
```
5. Create `.eslintrc.yml`
```yml
extends:
  - next
```
6. `yarn next lint`

**Comment by samsisle** [comment]
@kimbaudi Unable to replicate with your repro 🤔

**Comment by kimbaudi** [comment]
strange... I'm able to reproduce the error:

![Untitled](https://user-images.githubusercontent.com/1864092/131903293-3be73928-37b2-4de2-acb8-c59d1ac5cc0e.png)

I just did the following:
1. `git clone https://github.com/kimbaudi/test-next-app.git`
2. `cd test-next-app`
3. `npm install && npm run build`

I'm on Windows 10 using node v14.17.6 (LTS) and npm v 7.21.1.

Can anyone else check if they get the error by cloning my repo?

**Comment by kimbaudi** [comment]
@samsisle - your screenshot does not look like my repo at https://github.com/kimbaudi/test-next-app. My repo (and screenshot) has `_document.tsx` under `src/pages` folder. Your screenshot just has `_document.jsx` under `src` folder. Are you sure you are testing my repo?

**Comment by kimbaudi** [comment]
Let me try on my Ubuntu VM, maybe this issue only affects Windows users

**Comment by samsisle** [comment]
@kimbaudi That was for the other repro.

Here is yours →

<img width="2032" alt="CleanShot 2021-09-02 at 12 24 16@2x" src="https://user-images.githubusercontent.com/28912696/131904185-eae22ab8-e119-4614-abdf-47f9479d7f93.png">

**Comment by housseindjirdeh** [comment]
Thanks so much for the reproduction steps @kimbaudi and @rtritto. I also haven't been able to reproduce using either of your examples so this seems like it might just have to be a Windows issue

(Thanks @samsisle for trying them both out as well 🙏 )

**Comment by kimbaudi** [comment]
@housseindjirdeh @samsisle - I can confirm that this issue only affects Windows users. I just tried on my Ubuntu VM and am able to build without error.

**Comment by samsisle** [comment]
Can confirm as well that it's reproducible on Windows. Letting the Next.js team know!

**Comment by GrantGryczan** [comment]
Huh, why was this reopened?

**Comment by akd-io** [comment]
I'm hitting this error while trying to add Tailwind support to [Create Next Stack](https://www.create-next-stack.com).

I have ensured the import is not present in any other files. I'm using `pages/_document.tsx` but also hit it when renaming to `pages/_document.js`.

Node version: v16.3.0

Relevant dependencies:
```
  "dependencies": {
    "next": "11.1.2",
    "react": "17.0.2",
    "react-dom": "17.0.2",
    ...
  },
  "devDependencies": {
    "@types/react": "17.0.19",
    "eslint": "7.32.0",
    "eslint-config-next": "11.1.2",
    "typescript": "4.4.2",
    ...
  },
```

**Comment by akd-io** [comment]
My guess is, the change to test `page.startsWith('/_document')` instead of `name.startsWith('_document')` messes up on Windows because it uses `\` instead of `/`.

**Comment by housseindjirdeh** [comment]
This is fixed in #28745 and works in `11.1.3-canary.7`. Please upgrade to that version if you're experiencing this issue and if you're still seeing the same error, make sure to clear your cache (`next lint --no-cache`).

**Comment by darrylhuffman** [comment]
Getting this same issue in `11.1.3-canary.14` - am I missing something?

**Comment by realdocweird** [comment]
> Getting this same issue in `11.1.3-canary.14` - am I missing something?

same with 11.1.3-canary.15

**Comment by ivenuss** [comment]
Temporary fix for me was to downgrade `eslint-config-next` on version `11.1.0`

**Comment by housseindjirdeh** [comment]
@darrylhuffman @realdocweird Hmm are you using Windows and have both the `next` and `eslint-config-next` packages set to version `11.1.3-canary.7` or later? If so, could you also make sure the ESLint cache is cleared by running `next lint --no-cache`.

Let me know if you're still seeing the issue and I can help investigate why it's still happening

**Comment by realdocweird** [comment]
i've tryed the canary.7 and .15. I'm nooby i didn't understand very well how to `next lint --no-cache` xD. I've "solved" downgrading to next 10.2.3 but of course i will need to update asap

**Comment by kimbaudi** [comment]
TLDR; it works for me (on both v11.1.3-canary.18 or v11.1.2)

I no longer get this error in either v11.1.3-canary.18 or v11.1.2. I thought I would still get this error on v11.1.2, but I don't anymore. The important thing I did was run `next lint --no-cache`. I believe running this is what allowed me to switch from v11.1.3-canary.18 to v11.1.2 without running into this error.

Steps I took:
- I was originally on next v11.1.0 & eslint-config-next v11.1.0
- upgrade next & eslint-config-next to v11.1.3-canary.18
- run `next lint --no-cache`
- run `next build` (I no longer get the error)
- downgrade next & eslint-config-next to v11.1.2
- run `next build` (I no longer get the error)

**Comment by rogor** [comment]
It now works for me on next & eslint-config-next v11.1.3-canary.18

The problem I had was that I was running `next lint --no-cache` the wrong way.
If you have in your package.json a script ` "lint": "next lint"`, you cannot run `npm run lint --no-cache`, it won't work (the `--no-cache` isn't transferred to the command).
You have to change it in your package.json to ` "lint": "next lint --no-cache"` and then run `npm run lint`.

It's one of those silly mistakes you might do if you're not paying attention :p

**Comment by Dylan0916** [comment]
This issue occurred in `11.1.2`.

**Comment by ognitio-technologies** [comment]
> I get the same false positive when using npm `next build`.
> 
> All I did was update from Next v11.1.0 to Next 11.1.2 and update eslint-config-next from 11.1.0 to 11.1.2.
> 
> I'm on Windows 10, using node v16.7.0.
> Everything was fine with next 11.1.0

Im using the same version but still the same issues

**Comment by nbouvrette** [comment]
To everyone who still has the issue after upgrading to both Next.js and Eslint >= v11.1.3-canary.18 - run

`npx next lint --no-cache`

Also, I had run `npm install --force` when upgrading to

```
  "dependencies": {
    "next": "^11.1.3-canary.41",
  },
  "devDependencies": {
    "eslint-config-next": "^11.1.3-canary.41",
  }
```
